src/main/java/com/facebook/openwifirrm/ucentral/UCentralClient.java

```diff
@@ -195,27 +202,117 @@ public boolean login() {
 		}
 
 		// Parse access token from response
-		JSONObject respBody;
+		WebTokenResult token;
 		try {
-			respBody = new JSONObject(response.getBody());
-		} catch (JSONException e) {
+			token = gson.fromJson(response.getBody(), WebTokenResult.class);
+		} catch (JsonSyntaxException e) {
 			logger.error("Login failed: Unexpected response", e);
 			logger.debug("Response body: {}", response.getBody());
 			return false;
 		}
-		if (!respBody.has("access_token")) {
+		if (token.access_token == null || token.access_token.isEmpty()) {
 			logger.error("Login failed: Missing access token");
-			logger.debug("Response body: {}", respBody.toString());
+			logger.debug("Response body: {}", token.toString());
 			return false;
 		}
-		this.accessToken = respBody.getString("access_token");
+		this.accessToken = token;
 		logger.info("Login successful as user: {}", username);
-		logger.debug("Access token: {}", accessToken);
+		logger.debug("Access token: {}", accessToken.access_token);
+		logger.debug("Refresh token: {}", accessToken.refresh_token);
 
 		// Load system endpoints
 		return loadSystemEndpoints();
 	}
 
+	/**
+	 * Check if an access token is expired.
+	 *
+	 * @return true if the access token is expired
+	 */
+	private boolean isAccessTokenExpired() {
+		return accessToken.created + accessToken.idle_timeout <
+			Instant.now().getEpochSecond();
+	}
+
+	/**
+	 * Check if the refresh token is expired.
+	 *
+	 * @return true if the refresh token is expired
+	 */
+	private boolean isRefreshTokenExpired() {
+		return accessToken.created + TOKEN_REFRESH_WINDOW_S <
+			Instant.now().getEpochSecond();
+	}
+
+	/**
+	 * Get access token. If the refresh token is expired, login again. 
+	 * If the access token is expired, POST a WebTokenRefreshRequest to refresh token. 
+	 * Otherwise return the current access token.
+	 *
+	 * @return a valid access token ({@code WebTokenResult})
+	 */
+	private WebTokenResult getAccessToken() {
+		if (isRefreshTokenExpired()) {
+			logger.info("Refresh token is expired, login again");
+			if (login()) {
+				return accessToken;
+			}
+			return null;
+		} else if (isAccessTokenExpired()) {
+			logger.debug("Access token is expired, start refreshing a token.");
+			WebTokenResult refresh_token = refreshToken();
+			if (refresh_token != null) {
+				logger.debug("Successfully refresh token.");
+				return refresh_token;
+			}
+			logger.error(
+				"Fail to refresh token with access token: {}",
+				accessToken.access_token
+			);
+			return null;
+		}
+		return accessToken;
+	}
+
+	/**
+	 * POST a WebTokenRefreshRequest to refresh the access token.
+	 *
+	 * @return valid access token if success, otherwise return null.
+	 */
+	private WebTokenResult refreshToken() {
+		WebTokenRefreshRequest refreshRequest = new WebTokenRefreshRequest();
+		refreshRequest.userId = username;
+		refreshRequest.refreshToken = accessToken.refresh_token;
```

Comment:
just trying to understand, does this mean that the refresh token is only used to refresh an existing access token?

Reply:
Yes